### PR TITLE
feat: Recordings use events_summary to add extra data and improve performance

### PR DESCRIPTION
**Pull request by benjackwhite**
## Problem

Needs https://github.com/PostHog/posthog/pull/12815 which helps with tidying the tests up

We added the `events_summary` and extra materialised columns to allow us to do more list visualisation and eventually querying. Also it would allow us to load the metadata without having to decompress the entire snapshot list

## Changes

* Changes this list query to include:
  * Click count
  * Keypress count
  * First URL
  * Accurate duration (based on first_event_timestamp and last_event_timestamp)

Screenshot 2022-11-16 at 14 45 39

As well as allowing us to build the active/inactive playbar without needing to decompress any raw data.

## TODO
- [x] Fix tests
- [ ] Ensure that it is backwards compatible with non-events-summary recordings (self-hosted)

### CODE REVIEW

**Comment by benjackwhite** [comment]
@rcmarron  and @alexkim205 keen to see what you think of this. I went a little further to figure out how would we actually use this summary data and realised we potentially want a lot more of the info than we were originally thinking. This would make `events_summary` a bit bigger but _i think_ most of the heavy weight still comes from the DOM stuff so this would be overall safe if we long term wanted to move the full snapshot data elsewhere.

This PR itself probably wouldn't go in, but rather if we agree that the data store approach makes sense I will make a tiny PR to take care of the **write path** alone so that we are at least populating the data.

File: posthog/models/session_recording_event/sql.py [hunk 1]
```diff
@@ -26,20 +26,24 @@
 
 SESSION_RECORDING_EVENTS_MATERIALIZED_COLUMNS = """
     , has_full_snapshot Int8 MATERIALIZED JSONExtractBool(snapshot_data, 'has_full_snapshot') COMMENT 'column_materializer::has_full_snapshot'
+    , events_summary Array(String) MATERIALIZED JSONExtract(JSON_QUERY(snapshot_data, '$.events_summary[*]'), 'Array(String)') COMMENT 'column_materializer::events_summary'
+    , click_count Int8 MATERIALIZED length(arrayFilter((x) -> JSONExtractInt(x, 'type') = 3 AND JSONExtractInt(x, 'data', 'source') = 2 AND JSONExtractInt(x, 'data', 'type') = 2, events_summary))
+    , keypress_count Int8 MATERIALIZED length(arrayFilter((x) -> JSONExtractInt(x, 'type') = 3 AND JSONExtractInt(x, 'data', 'source') = 5, events_summary))
+    , first_event_timestamp DateTime64(3, 'UTC') MATERIALIZED toDateTime(arrayReduce('min', arrayMap((x) -> JSONExtractInt(x, 'timestamp'), events_summary)) / 1000)
+    , last_event_timestamp DateTime64(3, 'UTC') MATERIALIZED toDateTime(arrayReduce('max', arrayMap((x) -> JSONExtractInt(x, 'timestamp'), events_summary)) / 1000)
+    , urls Array(String) MATERIALIZED arrayFilter(x -> x != '', arrayMap((x) -> JSONExtractString(x, 'data', 'href'), events_summary))
```
Comment (rcmarron) [inline posthog/models/session_recording_event/sql.py:34]: I think this is going to pull the href from the meta event, but that doesn't fire on new URLs for SPAs.

File: posthog/models/session_recording_event/sql.py [hunk 2]
```diff
@@ -116,8 +120,8 @@
 
 INSERT_SESSION_RECORDING_EVENT_SQL = (
     lambda: f"""
-INSERT INTO {SESSION_RECORDING_EVENTS_DATA_TABLE()} (uuid, timestamp, team_id, distinct_id, session_id, window_id, snapshot_data, created_at, _timestamp, _offset)
-SELECT %(uuid)s, %(timestamp)s, %(team_id)s, %(distinct_id)s, %(session_id)s, %(window_id)s, %(snapshot_data)s, %(created_at)s, now(), 0
+INSERT INTO {SESSION_RECORDING_EVENTS_DATA_TABLE()} (uuid, timestamp, team_id, distinct_id, session_id, window_id, snapshot_data, events_summary, created_at, _timestamp, _offset)
+SELECT %(uuid)s, %(timestamp)s, %(team_id)s, %(distinct_id)s, %(session_id)s, %(window_id)s, %(snapshot_data)s, %(events_summary)s, %(created_at)s, now(), 0
```
Comment (rcmarron) [inline posthog/models/session_recording_event/sql.py:124]: I don't think we need this now that snapshot_data contains the event_summary

**Review by rcmarron — COMMENTED** [review]
I love it! I mainly looked at the ingestion/data storage side and have a couple of comments there.

Outside of those comments:
* What does it look like if we wanted to do an "active time" query in clickhouse? Could we finagle something there - maybe an approximation from the event summary? This is a common request from users, and I think this work gets us very close to having it
* I wonder if there are other bits of data worth extracting. For example, from the console log data, does it make sense to extract if the log is a warning/error etc? Feels like there's a good exercise of running through the rr-web [data payloads](https://github.com/rrweb-io/rrweb/blob/master/packages/rrweb/src/types.ts#L19) and making sure we're grabbing what we want (note, that GH link doesn't have the plugin payloads)

File: posthog/helpers/session_recording.py [hunk 3]
```diff
@@ -250,11 +334,38 @@ def get_active_segments_from_event_list(
...
+    'data' values are included as long as they are strings or numbers and not in the exclusion list to keep the payload minimal
+    """
+    events_summary = [
+        SessionRecordingEventSummary(
+            timestamp=event["timestamp"],
+            type=event["type"],
+            data={
+                key: value
+                for key, value in event.get("data", {}).items()
+                if type(value) in [str, int] and key not in EVENT_SUMMARY_DATA_EXCLUSIONS
```
[the remark this answers is not in the document]
Comment (benjackwhite) [reply posthog/helpers/session_recording.py:352]: That is some great input :D I had the `EVENT_SUMMARY_DATA_EXCLUSIONS` but maybe its better to make it an `INCLUSION` list instead. I'll make that change and propose it in a dedicated PR just for the generation of this info so we can review any missing data there.

File: posthog/models/session_recording_event/sql.py (hunk 1 above)
Comment (benjackwhite) [reply posthog/models/session_recording_event/sql.py:34]: Yeah I didn't realise we used the pageview custom event thing so I'll pull that out explicitly.

File: posthog/models/session_recording_event/sql.py (hunk 2 above)
Comment (benjackwhite) [reply posthog/models/session_recording_event/sql.py:124]: True. If we merge the write path ASAP then we might be able to figure out a point in time where we load everything via event_summary alone and then fallback to a different query for much older items. The migration here is gonna be interesting...

**Pull request merged**